d3-selection/index.d.ts

```diff
@@ -845,7 +878,7 @@ interface Selection<GElement extends BaseType, Datum, PElement extends BaseType,
     /**
      * Return the first (non-null) element in this selection. If the selection is empty, returns null.
      */
-    node(): GElement | null;
+    node(): GElement;
```

Comment:
The `this` context on a selection will [never be null](https://github.com/d3/d3-selection/blob/master/src/selection/each.js#L5). The callback provided to `.each` is invoked if and only if the node is truthy. No selection method will ever see a null value, making it somewhere between useless and counterproductive to include it in the selection's type parameter itself.

One further wrinkle: once [`keyof T`](https://github.com/Microsoft/TypeScript/pull/11929) lands, we'll want to extend `Selection.property` to use it for better type checking. At that point, including `null` in the selection type (as opposed to special-casing it) will interfere with that goal, since touching a property on `null` throws an error.

At least to me, this implies that semantically, a `Selection` is a container of non-null nodes, and it's completely safe, even in the presence of `strictNullChecks`, to omit the `| null` union arm.

The fact that `.node()` (and indeed, a number of other getter-style methods) return `null`, then, is a failure mode and no different from the fact that `document.querySelector()` or `document.getElementById()` can return `null`: the CSS query (or passed-in element) didn't exist. And that's fine! These things happen. If the user wants to kick the null-checking can to the runtime, they have the `!` operator:

```typescript
// Both of these queries should behave identically, since they actually *are* identical:
const node1 = document.querySelector('.known-to-exist')!;
const node2 = d3.select('.known-to-exist').node()!;
console.log(node1.textContent, node2.textContent);
```

This extends past `select` and into `filter`. It's very easy to write a buggy filter method that can empty out the selection for any number of reasons, and the way this PR introduces that type, it makes it very easy to accidentally suppress type checks that the compiler would have otherwise caught:

```typescript
const node = d3.selectall<HTMLDivElement>('div')
  .filter(() => false)
  .node();

console.log(node.textContent); // A crash not caught by the compiler.
```

The gist of my argument is that moving `| null` to the `Selection` type parameter ultimately means that, if the user provides a type parameter, they have to opt back in to null checks. That seems to go against the spirit of `strictNullChecks`. If they want to opt out of a null check, they can use the already-established `!` operator.